### DISCUSSION TITLE
Enhance Microstream REST Interface to use Java Stream API for Querying data

### DISCUSSION THREAD

**Discussion by dcoracle** [microstream-one/microstream#305] [Ideas]
Are there plans on introducing some sort of query language (i.e. Java Stream API) to navigate the Microstream Store?  Seems like it would be a worthwhile enhancement as it is difficult to navigate through millions of objects with the current interface.

**Comment by hg-ms** [comment]
We currently don’t plan to implement such a feature.
One reason is that the REST Interface directly access the (low level) persisted data instead of the objects in memory instances.

**Reply by dcoracle** [reply]
Are there plans to implement something like this for the objects in memory?  We've come up with a REPL-like interface via GroovyShell that can access these Root Instances which works for us.  However, it would be nice to have this native to the product.

**Reply by fh-ms** [reply]
Have you tried JShell? It is the REPL for Java.

**Reply by dcoracle** [reply]
A SO article (https://stackoverflow.com/questions/44929943/can-the-java-9-jshell-be-used-to-run-code-in-another-jvm) stated that JShell runs on its own JVM which does not work for us.  Implementing/enhancing something that attaches via Java Agent is possible but wanted to see if there was an easier way to directly run a Shell within the same ClassLoader as the Root Instances.  I did a quick POC/test using GroovyShell and it seems to work well.  It also supports Java Stream API which is what I wanted

**Reply by dcoracle** [reply]
Here is a very quick and dirty test of what I did:
```
    @Path("/microstream/streamApiQuery")
    @POST
    @Consumes(MediaType.TEXT_PLAIN)
    @Produces(MediaType.TEXT_PLAIN)
    public String streamApiQuery(String payload) {
        try {
            List<Person> personList = cacheProvider.getAllPeoplePojo();

            Binding binding = new Binding();
            binding.setVariable("personList", personList);
            GroovyShell shell = new GroovyShell(binding);
            
            String groovyScript = payload;
            ArrayList<Person> rtnArr = ((ArrayList<Person>) shell.evaluate(groovyScript));

            ObjectMapper mapper = new ObjectMapper();
            return mapper.writeValueAsString(rtnArr);

        }catch(Throwable e){
            e.printStackTrace();
        }

        return "[]";
    }
```

And to test:
```
curl -X POST -H "Content-Type: text/plain" "http://localhost:7001/cache/microstream/streamApiQuery" --data "personList.stream().filter(person -> person.getDisplayName().contains(\"Joe\")).toList();"
```
